README.md
[![license](https://img.shields.io/npm/l/ghost-town.svg?style=flat)](http://opensource.org/licenses/MIT) [![version](https://img.shields.io/npm/v/ghost-town.svg?style=flat)](https://www.npmjs.com/package/ghost-town) [![dependencies](https://img.shields.io/david/buzzvil/ghost-town.svg?style=flat)](https://david-dm.org/buzzvil/ghost-town)  
Simple queued & clustered PhantomJS processing. https://www.npmjs.com/package/ghost-town

*Now with 100% creepier dependencies! Check out Ghost Town 3's breaking changes in [CHANGELOG.md](CHANGELOG.md).*

---

Need highly scalable PhantomJS processing? Ghost Town makes it frighteningly easy! For example, on-demand page rendering, dispatched through Thrift:

    var town = require("ghost-town")();
    
    if (town.isMaster) {
        thrift.createServer(Renderer, {
            render: function (html, width, height, next) {
                town.queue({
                    html: html,
                    width: width,
                    height: height
                }, function (err, data) {
                    next(err, !err && new Buffer(data, "base64"));
                });
            }
        }).listen(1337);
    } else {
        town.on("queue", function (page, data, next) {
            // sequential page setup
            // page.property("viewportSize", ...)
            // page.property("customHeaders", ...)
            // page.property("onLoadFinished", ...)
            // page.property("content", ...)
            
            page.renderBase64("jpeg").then(function (res) {
                next(null, res);
            }).catch(next);
        });
    }

Ghost Town uses Node's Cluster API, so the master and worker share their code. On the master side, queue items and handle their results. On the worker side, process items and return their results.

Requires Node 4+ and PhantomJS 2.1+.

---

`town(options)`

* `phantomBinary`: String path to the PhantomJS executable. Default: Automatic via `$PATH`.
* `phantomFlags`: Object of strings to use for the PhantomJS options. Default: `{}`.
* `workerCount`: Number of workers to maintain. One or two per CPU is recommended. Default: `4`.
* `workerDeath`: Number of items to process before restarting a worker. Default: `25`.
* `workerShift`: Number of milliseconds to wait before restarting a worker. Default: `-1` (forever).
* `pageCount`: Number of pages to process at a time. If your processing is mostly asynchronous (vs. e.g. render blocked), increasing this is recommended. Default: `1`.
* `pageDeath`: Number of milliseconds to wait before before requeuing an item. If your processing is time-sensitive, decreasing this is recommended. Default: `30000`.
* `pageTries`: Number of times to retry items that have timed out. If your processing could fail forever, configuring this is recommended. Default: `-1` (unlimited).

Starts Ghost Town and returns a `Master` or a `Worker` instance exposing the following.

* `Master#isRunning` is set by `Master#start()` and `Master#stop()`.
* `Master#isMaster` and `Worker#isMaster` can be used to separate master- and worker-specific code.
* `Worker#phantom` is the PhantomJS wrapper object provided by [phantom](https://www.npmjs.com/package/phantom).

`Master#start()` and `Master#stop()`  
Starts or stops processing. These spawn or kill workers and PhantomJS processes, so they're useful for managing resource usage or gracefully shutting down Node.

`Master#queue(data, [asap], next)`  
Queue an item for processing by a worker. `data` is passed to `Worker!queue()`, and `next(err, data)` is called when complete. Optionally pass `true` to `asap` to prepend to the queue.

`Worker!queue(page, data, next)`  
Fired when a worker receives an item to process. `page` is the PhantomJS page, `data` is what was passed to `Master#queue()`, and `next(err, data)` passes it back.

---

© 2016 [Buzzvil](http://www.buzzvil.com), shared under the [MIT license](http://www.opensource.org/licenses/MIT).


CHANGELOG.md
3.0.0 / 2016-07-22
==================

Ghost Town 3 is a major release with breaking changes in every dependency. Please carefully review each before upgrading! For the `phantom` migration guide, see: https://github.com/amir20/phantomjs-node#migrating-from-10x

* Support Node 4+.
* Support PhantomJS 2.1+.
* Update `phantom` to `~2.1.12`.
* Remove the `phantomPort` option. `phantom` now uses pipes.

2.1.0 / 2015-02-27
==================

* Add the `workerShift` option. Use it if PhantomJS is occasionally never returning.

2.0.0 / 2015-02-05
==================

Ghost Town 2 is a major refactor release made up of several small but breaking changes. Please carefully review each before upgrading!

* Add tests!
* Add an optional `asap` argument to `Master#queue()`.
* Update `phantom` to `~0.7.2`. Now works with PhantomJS 2!
* Remove `phantom`'s default PhantomJS stdout and stderr logging. Listen to the `.stdout` and `.stderr` streams on the `Worker#phantom` object instead.
* Change the Ghost Town worker management algorithm so that it trusts workers less. May fix race conditions.
* Change the `phantomFlags` option to accept an object instead of an array. Separate the key and value so that `"--disk-cache=true"` becomes `"disk-cache": "true"`.
* Change some option defaults. `workerCount`: `os.cpus().length` to `4`. `pageDeath`: `120000` to `30000`. `workerDeath`: `20` to `25`.
* Rename `Master#running` to `Master#isRunning`.
* Rename all private properties to be private.


test.js
"use strict";

const async = require("async");
const child = require("child_process");
const cluster = require("cluster");
const expect = require("chai").expect;
const ghost = require("./");

let townSend;

if (cluster.isMaster) {
    cluster.setupMaster({
        exec: __filename,
    });
    
    cluster.on("online", function (worker) {
        worker.send(townSend || null);
    });
    
    afterEach(function (next) {
        let live = 1;
        
        function step () {
            if (!--live) {
                child.exec("killall phantomjs", next.bind(null, null));
            }
        }
        
        cluster.removeAllListeners("exit");
        
        for (let key in cluster.workers) {
            live++;
            cluster.workers[key].on("exit", step).kill();
        }
        
        step();
    });
    
    describe("Master", function () {
        this.timeout(5000);
        
        describe("constructor", function () {
            it("should support workerCount", function () {
                ghost({ workerCount: 8 });
                
                expect(Object.keys(cluster.workers)).to.have.length(8);
            });
            
            it("should support pageDeath", function (next) {
                townSend = {};
                let town = ghost({ pageDeath: 100 });
                
                async.waterfall([function (next) {
                    // Give Ghost Town a dummy task it can keep timing out on
                    // until phantom is warmed up and actually ready
                    town.queue(0, function () { next(); });
                }, function (next) {
                    // Prevent retries to get an immediate test result
                    town._itemRetries = 0;
                    town.queue(0, function (err) {
                        expect(err).to.be.null;
                        next();
                    });
                }, function (next) {
                    town.queue(100, function (err) {
                        expect(err).to.be.an.instanceof(Error);
                        next();
                    });
                }], next);
            });
            
            it("should support pageTries (-1)", function (next) {
                townSend = {};
                let town = ghost({ workerCount: 1, pageDeath: 0 });
                let keys = Object.keys(cluster.workers);
                let trys = 0;
                
                cluster.workers[keys[0]].on("message", function (msg) {
                    trys++;
                    
                    if (trys > 25) {
                        next();
                    }
                });
                
                town.queue(0, function () {
                    next(Error("shouldn't have completed"));
                });
            });

            it("should support pageTries (x)", function (next) {
                townSend = {};
                let town = ghost({ pageDeath: 0, pageTries: 42 });
                
                town.queue(0, function (err) {
                    expect(err).to.be.an.instanceof(Error);
                    
                    next();
                });
            });
        });
        
        describe("#start()", function () {
            it("should start Ghost Town", function () {
                let town = ghost();
                
                expect(town).to.respondTo("start");
                expect(town.isRunning).to.be.true;
            });
            
            it("should start all workers", function () {
                ghost({ workerCount: 5 });
                
                expect(Object.keys(cluster.workers)).to.have.length(5);
            });
            
            it("should restart Ghost Town and all workers", function () {
                let town = ghost({ workerCount: 5 });
                
                town.stop();
                town._workerCount = 11;
                town.start();
                
                expect(town.isRunning).to.be.true;
                expect(Object.keys(cluster.workers)).to.have.length(11);
            });
            
            it("should be idempotent", function () {
                let town = ghost({ workerCount: 5 });
                
                town.start();
                town.start();
                town.start();
                
                expect(Object.keys(cluster.workers)).to.have.length(5);
            });
        });
        
        describe("#stop()", function () {
            it("should stop Ghost Town", function () {
                let town = ghost();
                
                town.stop();
                
                expect(town).to.respondTo("stop");
                expect(town.isRunning).to.be.false;
            });
            
            it("should stop all workers", function () {
                let town = ghost();
                
                town.stop();
                
                expect(cluster.workers).to.be.empty;
            });
        });
        
        describe("#queue()", function () {
            it("should process items", function () {
                let town = ghost();
                
                expect(town).to.respondTo("queue");
            });
            
            it("should return results", function (next) {
                townSend = {};
                let town = ghost();
                
                town.queue(42, function (err, val) {
                    expect(err).to.be.null;
                    expect(val).to.equal(42);
                    
                    next();
                });
            });
            
            it("should support prepending", function () {
                let town = ghost();
                let curr = town._itemQueue;
                
                town.queue(0, function () {});
                town.queue(0, function () {});
                
                let orig = curr.slice();
                
                town.queue(0, town.queue);
                
                expect(curr).to.deep.equal([
                    orig[0],
                    orig[1],
                    curr[2]
                ]);
                
                town.queue(0, true, town.start);
                town.queue(0, false, town.stop);
                
                expect(curr).to.deep.equal([
                    curr[0],
                    orig[0],
                    orig[1],
                    curr[3],
                    curr[4]
                ]);
                
                expect(curr).to.have.deep.property("[0].done", town.start);
                expect(curr).to.have.deep.property("[3].done", town.queue);
                expect(curr).to.have.deep.property("[4].done", town.stop);
            });
        });
    });
    
    describe("Worker", function () {
        this.timeout(5000);
        
        describe("constructor", function () {
            // Pending release of https://github.com/amir20/phantomjs-node/pull/507
            it("should support phantomBinary");
            
            it("should support phantomFlags", function (next) {
                townSend = { _test: "pid", phantomFlags: { "disk-cache": "true" } };
                let town = ghost({ workerCount: 1 });
                
                town.queue(null, function (err, val) {
                    expect(err).to.be.null;
                    
                    child.exec("ps -p " + val + " -o command | sed 1d", function (err, out) {
                        expect(out).to.contain("--disk-cache=true");
                        
                        next();
                    });
                });
            });
            
            it("should support workerDeath", function (next) {
                townSend = { workerDeath: 10 };
                let town = ghost({ workerCount: 1 });
                let orig = Object.keys(cluster.workers);
                
                expect(orig).to.have.length(1);
                
                async.timesSeries(11, function (n, next) {
                    expect(cluster.workers).to.have.keys(orig);
                    town.queue(0, next);
                }, function () {
                    expect(cluster.workers).to.not.have.keys(orig);
                    
                    next();
                });
            });
            
            it("should support workerShift", function (next) {
                townSend = { workerShift: 10 };
                let town = ghost({ workerCount: 1 });
                let orig = Object.keys(cluster.workers);
                
                expect(orig).to.have.length(1);
                
                setTimeout(function () {
                    expect(cluster.workers).to.not.have.keys(orig);
                    
                    next();
                }, 2000);
            });
            
            it("should support pageCount", function (next) {
                townSend = { pageCount: 3 };
                let town = ghost({ workerCount: 1 });
                
                town.queue(0, function () { setImmediate(function () {
                    town.queue(0);
                    town.queue(0);
                    town.queue(0);
                    
                    expect(town._itemQueue).to.be.empty;
                    
                    town.queue(0);
                    town.queue(0);
                    
                    expect(town._itemQueue).to.have.length(2);
                    
                    next();
                }); });
            });
        });
        
        describe("!queue", function () {
            it("should pass arguments", function (next) {
                townSend = { _test: "passArgs" };
                let town = ghost();
                
                town.queue(42, next);
            });
            
            it("should have an idempotent callback", function (next) {
                townSend = { _test: "passOnce" };
                let town = ghost({ workerCount: 3 });
                
                town.queue(0, function (err, val) {
                    setTimeout(function () {
                        expect(town._workerQueue).to.have.length(3);
                        
                        next();
                    }, 500);
                });
            });
        });
    });
    
    describe("Ghost Town", function () {
        this.timeout(25000);
        
        it("should handle heavy loads", function (next) {
            townSend = {};
            let town = ghost();
            
            async.times(500, function (n, next) {
                town.queue(n / 100, function (err, val) {
                    if (err) {
                        next(new Error("shouldn't have errored: " + err));
                    } else if (val !== n / 100) {
                        next(new Error("returned the wrong data"));
                    } else {
                        next();
                    }
                });
            }, next);
        });
        
        it("should handle many timeouts", function (next) {
            townSend = {};
            let town = ghost({ pageDeath: 100, pageTries: 0 });
            
            async.times(100, function (n, next) {
                town.queue(n % 1 ? (n / 100) : 101, function (err, val) {
                    if (n % 1) {
                        if (err) {
                            next(new Error("shouldn't have timed out"));
                        } else if (val !== n / 100) {
                            next(new Error("returned the wrong data"));
                        } else {
                            next();
                        }
                    } else if (!err) {
                        next(new Error("should have timed out"));
                    } else {
                        next();
                    }
                });
            }, next);
        });
    });
} else {
    const townTest = {
        pid: function (town, page, data, next) {
            next(null, town.phantom.process.pid);
        },
        passArgs: function (town, page, data, next) {
            expect(page).to.be.an("object");
            expect(page.renderBase64).to.be.a("function");
            
            expect(data).to.equal(42);
            
            expect(next).to.be.a("function");
            
            next();
        },
        passOnce: function (town, page, data, next) {
            next();
            next();
            next();
        },
    };
    
    process.once("message", function (opts) {
        const town = ghost(opts).on("queue", function (page, data, next) {
            if (opts._test) {
                try {
                    townTest[opts._test](town, page, data, next);
                } catch (err) {
                    next(err);
                }
            } else {
                setTimeout(next, data, null, data);
            }
        });
    });
}


index.js
"use strict";

const cluster = require("cluster");
const events = require("events");
const phantom = require("phantom");

function is (type, val, def) {
    return val !== null && typeof val === type ? val : def;
}

class Master extends events.EventEmitter {
    constructor (opts) {
        opts = is("object", opts, {});
        
        super();
        
        this.isMaster = true;
        this.isRunning = false;
        
        this._workerCount = is("number", opts.workerCount, 4);
        this._workerQueue = [];
        
        this._itemTimeout = is("number", opts.pageDeath, 30000);
        this._itemRetries = is("number", opts.pageTries, -1);
        this._itemClicker = 0;
        this._itemQueue = [];
        this._items = {};
        
        cluster.on("exit", this._onExit.bind(this));
        
        this.start();
    }
    
    _onMessage (msg) {
        if (is("object", msg, {}).ghost !== "town") {
            return;
        }
        
        const item = this._items[msg.id];
        
        if (item) {
            delete this._items[msg.id];
            clearTimeout(item.timeout);
            item.done(msg.err, msg.data);
        }
        
        this._workerQueue.push(cluster.workers[msg.worker]);
        this._process();
    }
    
    _onTimeout (item) {
        delete this._items[item.id];
        
        if (item.retries === this._itemRetries) {
            item.done(new Error("[ghost-town] max pageTries"));
        } else {
            this.queue(item.data, true, item.done, item.retries + 1);
        }
    }
    
    _onExit (worker) {
        for (let id in this._items) {
            const item = this._items[id];
            
            if (item.worker === worker) {
                delete this._items[id];
                clearTimeout(item.timeout);
                this.queue(item.data, true, item.done, item.retries);
            }
        }
        
        if (this.isRunning) {
            cluster.fork().on("message", this._onMessage.bind(this));
        }
    }
    
    start () {
        if (this.isRunning) {
            return;
        }
        
        this.isRunning = true;
        
        for (let i = this._workerCount; i--;) {
            this._onExit();
        }
    }
    
    stop () {
        this.isRunning = false;
        
        for (let key in cluster.workers) {
            cluster.workers[key].kill();
        }
    }
    
    queue (data, asap, next, tries) {
        const item = {
            id: this._itemClicker++,
            timeout: -1,
            retries: tries || 0,
            data: data,
            done: next || asap,
        };
        
        this._itemQueue[next && asap ? "unshift" : "push"](item);
        this._process();
    }
    
    _process () {
        while (this._workerQueue.length && this._itemQueue.length) {
            const worker = this._workerQueue.shift();
            
            if (!worker || !worker.process.connected) {
                continue;
            }
            
            const item = this._itemQueue.shift();
            
            item.worker = worker;
            item.timeout = setTimeout(this._onTimeout.bind(this, item), this._itemTimeout);
            this._items[item.id] = item;
            
            worker.send({
                ghost: "town",
                id: item.id,
                data: item.data,
            });
        }
    }
}

class Worker extends events.EventEmitter {
    constructor (opts) {
        opts = is("object", opts, {});
        
        super();
        
        this.isMaster = false;
        
        this._workerDeath = is("number", opts.workerDeath, 25);
        this._workerShift = is("number", opts.workerShift, -1);
        
        this._pageCount = is("number", opts.pageCount, 1);
        this._pageClicker = 0;
        this._pages = {};
        
        const flagArr = [];
        const flagObj = is("object", opts.phantomFlags, {});
        
        for (let key in flagObj) {
            flagArr.push("--" + key + "=" + flagObj[key]);
        }
        
        phantom.create(flagArr, {
            phantomPath: opts.phantomBinary,
        }).then((proc) => {
            this.phantom = proc;
            
            for (let i = this._pageCount; i--;) {
                process.send({
                    ghost: "town",
                    worker: cluster.worker.id,
                });
            }
        });
        
        process.on("message", this._onMessage.bind(this));
        
        if (this._workerShift !== -1) {
            setTimeout(this._exit.bind(this), this._workerShift);
        }
    }
    
    _onMessage (msg) {
        if (is("object", msg, {}).ghost !== "town") {
            return;
        }
        
        this.phantom.createPage().then((page) => {
            this._pageClicker++;
            this._pages[msg.id] = page;
            this.emit("queue", page, msg.data, this._done.bind(this, msg.id));
        });
    }
    
    _done (id, err, data) {
        if (!this._pages[id]) {
            return;
        }
        
        this._pages[id].close();
        delete this._pages[id];
        
        process.send({
            ghost: "town",
            worker: cluster.worker.id,
            id: id,
            err: err,
            data: data,
        });
        
        if (this._pageClicker >= this._workerDeath) {
            this._exit();
        }
    }
    
    _exit () {
        this.phantom.process.on("exit", process.exit);
        this.phantom.exit();
    }
}

module.exports = (opts) => {
    return cluster.isMaster ? new Master(opts) : new Worker(opts);
};
